### DISCUSSION TITLE
Support for anchor positioning

### DISCUSSION THREAD

**Discussion by holvi-mikael** [zauberzeug/nicegui#3398] [Ideas / Feature Requests]
Hello everyone

## Introduction

Now that the anchor support [is out](https://developer.chrome.com/blog/anchor-positioning-api) on Chrome and Edge, and while we wait Safari and others, would you be open to start discussing how specific support for it could work in NiceGUI?

As a quick summary, anchor positioning lets you place or size an element relative to another element in a way that is exact but dynamic. As such, it provides us with another option for layout, an option that is very much "put this there" instead of the "document flow" of the traditional CSS.

A simple example of using anchor positioning could look like this:

```css
#reference {
    anchor-name: --reference;
}

#anchored-left-to-right {
    left: anchor(--reference right)
}
```

Since it is just CSS, it is already fully available in NiceGUI via `.style`. Unfortunately, CSS still has its quirks with expecting the container to be positioned, centering etc., so even simple use cases take too much thinking and hand-crafting.

NiceGUI could have a nice pythonic anchor positioning syntax, and do all that thinking and hand-crafting for us. Also, with Python attributes, we would get the benefit of IDE syntax completion.

## What it could look like

To me, anchor positioning would be an easier layout option for someone new to HTML/CSS. To demonstrate this, below is a basic UI layout, created in NiceGUI with anchor positioning CSS styles (and a little Javascript support to make entering and dynamic sizing easier).

![image](https://github.com/user-attachments/assets/ac6a45db-fcb7-4d28-8a05-80bed1e2570a)

Top-level component layout shows some basic features:

- Basic properties like top, left, center_x that are used to set dynamic anchors, like in `top = top_bar.bottom`.
- Support for calculations that are dynamically evaluated as layout changes, not only when first set in Python code, like `width = max(parent().width/3, 200))`.
- Convenience reference to `parent()` element.
- Convenience methods to anchor several edges at once, like `dock_top()`.
- Convenience methods to support dynamic size based on absolutely-positioned contents, like `fit_height()`.
- Meanwhile, the contents of the left panel is just a regular relatively positioned NiceGUI list component, driving the dynamic width of the panel in the normal way, demonstrating that anchor positioning can be mixed with flow positioning.

![image](https://github.com/user-attachments/assets/193050e3-ae94-4433-999f-81231a9eef8c)

Contents of the top bar show some more convenience functions for centering things, like `left_center()`.

![image](https://github.com/user-attachments/assets/9a22aee1-4d7d-40dc-9759-51e2bc7ff657)

Contents of the search box show convenience function `below()`, using another convenience function to refer to `previous()` sibling without having to set an explicit variable to reference the source of the anchor:

![image](https://github.com/user-attachments/assets/3f18bdb6-2712-41be-a9d0-cd50b66d5979)

Since these are still just CSS values, `transition` can be used for small animations:

https://github.com/user-attachments/assets/ff7c2401-8b15-4dd0-9f9f-e7e1146c0253

As a minor note, positioning the callouts in the pictures above was very easy with anchor positioning.

## Back to the questions

- Do you feel that anchor positioning would be an interesting area to explore while we wait for the full browser support?
- Do you see the value of adding explicit properties to make anchor positioning easy?
- Any thoughts on the actual API?

**Comment by falkoschindler** [comment]
Hi @holvi-mikael, thanks for the suggestion! It might be an interesting feature.

But it also sounds rather challenging to find an nice and intuitive API. You current suggestion mixes with the existing API of `ui.element` and probably needs some separation, e.g. like we did for `.tailwind`.

Let's hear what others think about it before jumping into the details.

**Comment by v479038280** [comment]
Such way is more convenient than the   `.tailwind`.  With the `.classes()`,the` .tailwind` is not very useful, if  the `.classes()` support  autocomplete ,it would be better

**Reply by python-and-novella** [reply]
It is a way to locate element, Do you mean this?
https://github.com/zauberzeug/nicegui/discussions/117

**Reply by v479038280** [reply]
"Such way"  I mean something like  [this](https://private-user-images.githubusercontent.com/62882723/352185706-193050e3-ae94-4433-999f-81231a9eef8c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI1MzA3ODksIm5iZiI6MTcyMjUzMDQ4OSwicGF0aCI6Ii82Mjg4MjcyMy8zNTIxODU3MDYtMTkzMDUwZTMtYWU5NC00NDMzLTk5OWYtODEyMzFhOWVlZjhjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAxVDE2NDEyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0ZDc3ZGFhMTA1NmUxMmJlMGU1YjMzMGVmYzQ5NjgyZjY3ODcwYTg5OTcyODU1ZWJmOTg4ZTNiNDRlYTZhODcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Eik2jRti7iWp0As6A5YOBmlsW9P0y5qaAcErn5_Ct1I), method or attribute like `.top()`,`.center()`

**Comment by mikaelho** [comment]
Thanks for the comments, much appreciated.

To avoid spamming the `ui.element` namespace, there could be just a single `.at()` function used for anchor positioning:

<img width="772" alt="image" src="https://github.com/user-attachments/assets/2ab7bf10-14bc-4418-9ed7-e8d3c5dc1387">

All the variables and functions like `bottom()`, `below()`, `fit_height`, and `previous` could live in a specific namespace, e.g. `nicegui.anchor`.

(I also played around with overriding `__matmul__` for something like `top=top_bar@bottom`, but that just looks too weird, no matter how apt.)

**Comment by mikaelho** [comment]
And another option, aligned with the `.tailwind` approach. It does seem the most concise option, even though, like `.tailwind`, it requires some extra awareness of whether you are riffing off the element or the anchor object. This option is very nice in terms of how it supports autocompletion.

<img width="891" alt="image" src="https://github.com/user-attachments/assets/df42d125-1ddc-4199-890a-334118151331">

**Comment by mikaelho** [comment]
I now have a first working implementation of anchor positioning in https://github.com/zauberzeug/nicegui/pull/3470.

There is a demo layout that looks like this:

<kbd>
<img src="https://github.com/user-attachments/assets/c0928ddf-2aad-4e01-83a1-138a2a3b6fa3">
</kbd>

... and was created like this:

```python
    with ui.element().classes("w-full h-full"):
        with ui.element().classes("bg-primary").at.top.fit_height as top_bar:
            ui.icon("home", size="lg", color="white").at.top_left
            ui.label("SAMPLE APP").style("margin: 10px").at.center.tailwind.text_color("white")
            ui.button(icon="menu").at.top_right
        with ui.element().at.bottom_left.a(top=bottom(top_bar)).classes("bg-primary") as drawer:
            add_left_drawer_items()
        with ui.element().at.bottom_right.a(top=bottom(top_bar), left=right(drawer)) as content_area:
            with ui.element().at.center.fit_height.a(width=max(width(parent)/3, 200)):
                input_with_id(placeholder="Username").at.top
                input_with_id(placeholder="Password").at.below(previous).a(width=width(previous))
                ui.button("Login").at.below(previous)
        ui.button(icon="add").at.bottom_right.props("fab")
```
